core/src/main/java/org/elasticsearch/index/query/MultiMatchQueryBuilder.java

```diff
@@ -295,82 +411,238 @@ public MultiMatchQueryBuilder cutoffFrequency(float cutoff) {
         }
-
         if (tieBreaker != null) {
             builder.field("tie_breaker", tieBreaker);
         }
-
-        if (lenient != null) {
-            builder.field("lenient", lenient);
+        builder.field("lenient", lenient);
+        if (cutoffFrequency != null) {
+            builder.field("cutoff_frequency", cutoffFrequency);
         }
+        builder.field("zero_terms_query", zeroTermsQuery.toString());
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
 
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        MultiMatchQuery multiMatchQuery = new MultiMatchQuery(context);
+        if (analyzer != null) {
+            if (context.analysisService().analyzer(analyzer) == null) {
+                throw new QueryShardException(context, "[" + NAME + "] analyzer [" + analyzer + "] not found");
+            }
+            multiMatchQuery.setAnalyzer(analyzer);
+        }
+        multiMatchQuery.setPhraseSlop(slop);
+        if (fuzziness != null) {
+            multiMatchQuery.setFuzziness(fuzziness);
+        }
+        multiMatchQuery.setFuzzyPrefixLength(prefixLength);
+        multiMatchQuery.setMaxExpansions(maxExpansions);
+        multiMatchQuery.setOccur(operator.toBooleanClauseOccur());
+        if (fuzzyRewrite != null) {
+            multiMatchQuery.setFuzzyRewriteMethod(QueryParsers.parseRewriteMethod(context.parseFieldMatcher(), fuzzyRewrite, null));
+        }
+        if (tieBreaker != null) {
+            multiMatchQuery.setTieBreaker(tieBreaker);
+        }
         if (cutoffFrequency != null) {
-            builder.field("cutoff_frequency", cutoffFrequency);
+            multiMatchQuery.setCommonTermsCutoff(cutoffFrequency);
+        }
+        multiMatchQuery.setLenient(lenient);
+        multiMatchQuery.setZeroTermsQuery(zeroTermsQuery);
+
+        if (useDisMax != null) { // backwards foobar
+            boolean typeUsesDismax = type.tieBreaker() != 1.0f;
+            if (typeUsesDismax != useDisMax) {
+                if (useDisMax && tieBreaker == null) {
+                    multiMatchQuery.setTieBreaker(0.0f);
+                } else {
+                    multiMatchQuery.setTieBreaker(1.0f);
+                }
+            }
         }
 
-        if (zeroTermsQuery != null) {
-            builder.field("zero_terms_query", zeroTermsQuery.toString());
+        Map<String, Float> newFieldsBoosts = handleFieldsMatchPattern(context.mapperService(), fieldsBoosts);
+
+        Query query = multiMatchQuery.parse(type, newFieldsBoosts, value, minimumShouldMatch);
+        if (query == null) {
+            return null;
         }
+        return query;
+    }
 
-        printBoostAndQueryName(builder);
+    @Override
+    protected void setFinalBoost(Query query) {
+        query.setBoost(boost * query.getBoost());
+    }
```

Comment:
I think the base method in AbstractQueryBuilder has already extensive javadocs, maybe we can add a simple comment here explaining that the query might get a boost coming out of field boosts parsing, which we need to combine with the boost assigned to the query as part of the query dsl.

That said, I don't see this same behaviour in master, it feels like it is the right behaviour especially looking at what query_string does, but on master we actually replace the parsed boost with the boost assigned to the query in the query DSL. 

Maybe open a separate issue and mark it fixed together with this PR so we at least keep track of this small change?